Add Support for new Rumor Message Data

Kaz-ookid opened the pull request:
This PR introduces support for the new `Rumor` message data type, in anticipation of the backend's forthcoming rework of network protocols to incorporate gossiping mechanisms. It specifically addresses #1789, that adds this new message type.

### Context
With the backend starting on the implementation of the gossiping protocol and the creation of new message types (refer to issue #1794 and PR #1789), we need to anticipate support a new message type designed for the propagation of rumors, so that it will be ready as soon as they start using it. 

### Implementation Details
The `Rumor` message type encapsulates:
- A unique `rumor_id` for quick identification of previously received rumors.
- A `sender_id`.
- A key-value pairing of channels to an array of messages.

Validation functions have been added to `MessageValidator.kt` to fit the needs.

<details>
<summary>JSON Structure of Rumor Message</summary>

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/rumor.json",
  "description": "Rumor that should be spread across the network",
  "type": "object",
  "additionalProperties": false,
  "properties": {
    "method": {
      "description": "[String] operation to be performed by the query",
      "const": "rumor"
    },
    "params": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "sender_id" : {
          "description": "[String] publish key of the sender's server",
          "type": "string"
        },
        "rumor_id": {
          "description": "[Integer] ID of the rumor",
          "type": "integer"
        },
        "rumors": {
          "description": "[Array] Key-value of channels and messages per channel",
          "type": "array",
          "items": {
            "$ref": "../../answer/result/messages_by_channel.json"
          },
          "minItems": 1
        }
      },
      "required": ["sender_id","rumor_id", "rumors"]
    },
    "jsonrpc": {
      "$comment": "Defined by the parent, but needed here for the validation"
    },
    "id": {
      "type": "integer"
    }
  },
  "required": ["method", "params", "id", "jsonrpc"]
}
```
</details>

**Note:**  modifications to a few other files are also to come (`utility.handler.data`, `DataRegistryModule`)
Tests are to come, after validation of the suggested structure.


### EDIT

Rumors will not be handled by clients, yet. This means this PR is not as useful as it should have been. Although, the Rumor message data and tests will still be merged so that further use of Rumors can be considered.

matteosz commented on fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/MessageValidator.kt:
```diff
@@ -231,6 +242,27 @@ object MessageValidator {
+     */
+    fun validMessage(message: Any): MessageValidatorBuilder {
+      message as Map<*, *>
+      isNotEmptyBase64(message["data"] as String?, "data")
+      isNotEmptyBase64(message["sender"] as String?, "sender")
+      isNotEmptyBase64(message["signature"] as String?, "signature")
+      isNotEmptyBase64(message["message_id"] as String?, "message_id")
```
I see a possible optimization to the code here: what about extending the `isNotEmptyBase64` method to support as input a variable number of messages? It would reduce a bit the code repetition here and in the future

Kaz-ookid replied:
makes sense!

matteosz reviewed (commented):
I see it's still a wip but wanted to add a small comment in the meanwhile

quadcopterman commented on fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/MessageValidator.kt:
```diff
@@ -167,6 +179,17 @@ object MessageValidator {
+     * @param input the string to check
+     * @param field name of the field (to print in case of error)
+     */
+    fun isNotNegative(input: Int, field: String): MessageValidatorBuilder {
```
The doc says that input is a string but it is defined as an Int. Is there a specific reason for that ?

Kaz-ookid replied:
yeah my bad, thanks!

quadcopterman reviewed (commented):
Just one comment, other than that it looks good to me

matteosz commented on fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/MessageValidator.kt:
```diff
@@ -231,6 +254,28 @@ object MessageValidator {
+     * @param message the message to check, expected to be a map with string keys and optional
+     *   string values
+     */
+    fun validMessage(message: Any): MessageValidatorBuilder {
+      message as Map<*, *>
+
+      val data = message["data"] as String?
+      val sender = message["sender"] as String?
+      val signature = message["signature"] as String?
+      val messageId = message["message_id"] as String?
```
If you manage to cast that `Any `to `MessageGeneral `(or create another class for it), then you'd already have the fields in the message object without needing to access it as a map.

Kaz-ookid commented:
Ready for final review! 

- Upgraded tests (+5% coverage and actually testing the validations)
- Changed the `messages` type (in `Rumor`) from ` List<Map<String, List<Any>>>` to `Map<String, List<MessageGeneral>>`
- Adapted tests for this change above
- Addressed other comments

matteosz reviewed (approved):
LGTM!
